lib/gpt/algorithms/inverter/multi_grid.py

```diff
@@ -136,6 +136,15 @@ def __init__(self, mat_f, params):
                 ]
             self.distribution[lvl](self.basis[lvl][0 : self.nb[lvl]])
 
+        # open bc for basis vectors if finest level matrix is open bc
+        # the way I retrieve this info is rather ugly -> TODO?
+        if self.mat[self.finest].params["boundary_phases"][-1] == 0.0:
```

Comment:
Agree, we need a better way to retrieve this info but also open boundaries should not be hard-coded in a solver.  We need to understand the abstract idea and implement it.  The idea is likely just that the basis needs to be created differently, so it needs to be located with the basis creation?

Reply:
This is the place where the basis vectors are filled with random numbers initially. On top of that open bc requires fermion fields to be zero at the temporal boundaries, which is what I do here. It's really that simple and I don't think there's a more abstract idea behind this.
The basis creation is located in the same file a few lines below this. However, moving this there would not improve on your points of criticism in my opinion.

Reply:
1) The solver should not know about parameters of the matrix it needs to invert.  Even assuming the existence of the params is poor design.

2) The solver *should* however consider itself with its own parameters, one of which is that the MG needs a prescription to go from coarse to fine and back or equivalently it needs a blocked basis.  The boundary phases are just a very special case of the abstract idea of how to create the blocked basis.

You can leave this to me, I will think about a solution and implement it.